File: collatz_hash.py
#!/usr/bin/env python3
"""
Collatz Hash - Pure math hashing using the Collatz conjecture.

No external crypto libraries. Just arithmetic and bit operations.

Output: Hexadecimal hash of the Collatz sequence sum.
"""

# Configuration
MULTIPLIER = 100_000_000_000_000  # 100 trillion
NUM_LEVELS = 15                    # Fake branch levels


def text_to_number(text, multiplier=MULTIPLIER):
    """
    Encode text to a number using base-128.
    
    Base-128 matches ASCII printable range (0-127 = 7 bits per char).
    """
    result = 0
    for char in text:
        result = (result << 7) + (ord(char) & 0x7F)
    result *= multiplier
    if not (result & 1):
        result = (result << 1) | 1
    return result


def collatz_sequence(n):
    """
    Generate Collatz sequence using bit operations for speed.
    """
    sequence = [n]
    while n != 1:
        if n & 1:  # Odd
            n = 3 * n + 1
        else:      # Even
            n = n >> 1
        sequence.append(n)
    return sequence


def collatz_hash(text, hex_length=None, bit_width=512):
    """
    Generate a Collatz-based hash of the input text.
    
    Uses 2D vector folding:
    1. Concatenate all sequence values into one large binary number
    2. Split into rows of 'bit_width' bits (e.g., 512)
    3. Pad last row if needed
    4. XOR/ADD all rows together to produce fixed-width output
    
    Args:
        text: Input string to hash
        hex_length: Optional - override output length in hex chars
        bit_width: Width of each row (default 512 bits)
    
    Returns:
        Hexadecimal hash string (fixed length based on bit_width)
    """
    # Encode text to number
    encoded = text_to_number(text)
    
    # Generate Collatz sequence
    sequence = collatz_sequence(encoded)
    
    # Concatenate all sequence values into one giant number
    concat_bits = 0
    total_bits = 0
    for val in sequence:
        val_bits = val.bit_length() or 1  # At least 1 bit
        concat_bits = (concat_bits << val_bits) | val
        total_bits += val_bits
    
    # Split into rows of bit_width
    mask = (1 << bit_width) - 1  # Mask for bit_width bits
    rows = []
    
    remaining = concat_bits
    while remaining > 0:
        row = remaining & mask  # Take lowest bit_width bits
        rows.append(row)
        remaining = remaining >> bit_width
    
    # If no rows (shouldn't happen), create one with zeros
    if not rows:
        rows = [0]
    
    # Pad the last row is already handled (it's just smaller, but that's fine)
    # Actually, we want to ensure it's treated as full width - it already is via the mask
    
    # Fold all rows together using directional folding
    # Even rows: add top-to-bottom (normal)
    # Odd rows: add bottom-to-top (bit-reversed within the row)
    
    def reverse_bits(n, width):
        """Reverse the bits of n within the given width."""
        result = 0
        for _ in range(width):
            result = (result << 1) | (n & 1)
            n >>= 1
        return result
    
    result = 0
    for i, row in enumerate(rows):
        if i % 2 == 0:
            # Even rows: XOR top-to-bottom (normal direction)
            result = result ^ row
        else:
            # Odd rows: ADD bottom-to-top (reversed bits)
            reversed_row = reverse_bits(row, bit_width)
            result = (result + reversed_row) & mask
    
    # Additional mixing: rotate and XOR with weighted sum
    weighted_sum = sum(val * (i + 1) for i, val in enumerate(sequence)) & mask
    result = result ^ weighted_sum
    
    # Convert to hexadecimal
    hex_hash = format(result, 'x')
    
    # Ensure fixed length (bit_width / 4 hex chars)
    target_len = hex_length if hex_length else (bit_width // 4)
    hex_hash = hex_hash.zfill(target_len)
    
    # Truncate if somehow longer
    if len(hex_hash) > target_len:
        hex_hash = hex_hash[:target_len]
    
    return hex_hash


def collatz_hash_256(text):
    """Return a 256-bit (64 hex char) hash."""
    return collatz_hash(text, bit_width=256)


def collatz_hash_512(text):
    """Return a 512-bit (128 hex char) hash."""
    return collatz_hash(text, bit_width=512)


def collatz_hash_full(text):
    """Return 512-bit hash (default)."""
    return collatz_hash(text, bit_width=512)


# Demo
if __name__ == "__main__":
    test_inputs = [
        'hello',
        'Hello',
        'password',
        'password1',
        'Noobas123!@#',
        'The quick brown fox jumps over the lazy dog',
    ]
    
    print("=" * 80)
    print("COLLATZ HASH - Pure Math, No Crypto Libraries")
    print("=" * 80)
    print()
    
    print("FULL HASH (variable length, zero collisions):")
    print("-" * 80)
    for text in test_inputs:
        h = collatz_hash_full(text)
        print(f'"{text}"')
        print(f'  Length: {len(h)} hex chars ({len(h)*4} bits)')
        print(f'  Hash: {h[:64]}{"..." if len(h) > 64 else ""}')
        print()
    
    print()
    print("256-BIT HASH (64 hex chars, like SHA-256):")
    print("-" * 80)
    for text in test_inputs:
        h = collatz_hash_256(text)
        print(f'"{text}"')
        print(f'  {h}')
        print()
    
    print()
    print("512-BIT HASH (128 hex chars, like SHA-512):")
    print("-" * 80)
    for text in test_inputs[:3]:  # Just first 3 for brevity
        h = collatz_hash_512(text)
        print(f'"{text}"')
        print(f'  {h}')
        print()
    
    print()
    print("=" * 80)
    print("COMPARISON: Same input = Same hash")
    print("=" * 80)
    h1 = collatz_hash_256('test')
    h2 = collatz_hash_256('test')
    print(f'collatz_hash_256("test") run 1: {h1}')
    print(f'collatz_hash_256("test") run 2: {h2}')
    print(f'Match: {h1 == h2}')
    print()
    
    print("=" * 80)
    print("AVALANCHE EFFECT: Small change = Completely different hash")
    print("=" * 80)
    for pair in [('hello', 'hallo'), ('abc', 'abd'), ('test', 'Test')]:
        h1 = collatz_hash_256(pair[0])
        h2 = collatz_hash_256(pair[1])
        # Count differing characters
        diff = sum(1 for a, b in zip(h1, h2) if a != b)
        print(f'"{pair[0]}" vs "{pair[1]}": {diff}/64 chars differ ({diff/64*100:.1f}%)')

File: collatz_hash.cpp
/**
 * Collatz Hash - C++ implementation for maximum speed.
 * Pure math hashing using the Collatz conjecture. No external crypto libs.
 *
 * Requires: GMP (GNU Multiple Precision) - brew install gmp
 * Build: g++ -std=c++17 -O3 -o collatz_hash collatz_hash.cpp -lgmp -lgmpxx
 */
#include <chrono>
#include <iostream>
#include <string>
#include <gmpxx.h>

constexpr long MULTIPLIER = 100'000'000'000'000L;  // 100 trillion

mpz_class text_to_number(const std::string& text) {
    mpz_class result = 0;
    for (unsigned char c : text) {
        result = (result << 7) + (c & 0x7F);
    }
    result *= MULTIPLIER;
    if (result % 2 == 0) {
        result = (result << 1) | 1;
    }
    return result;
}

void collatz_sequence(mpz_class n, std::vector<mpz_class>& out) {
    out.clear();
    out.push_back(n);
    while (n != 1) {
        if (mpz_odd_p(n.get_mpz_t())) {
            n = 3 * n + 1;
        } else {
            n = n >> 1;
        }
        out.push_back(n);
    }
}

mpz_class reverse_bits(mpz_class n, int width) {
    mpz_class result = 0;
    for (int i = 0; i < width; ++i) {
        result = (result << 1) | (n & 1);
        n >>= 1;
    }
    return result;
}

std::string collatz_hash(const std::string& text, int bit_width = 512) {
    mpz_class encoded = text_to_number(text);

    std::vector<mpz_class> sequence;
    collatz_sequence(encoded, sequence);

    // Concatenate all sequence values into one giant number
    mpz_class concat_bits = 0;
    for (const auto& val : sequence) {
        size_t val_bits = mpz_sizeinbase(val.get_mpz_t(), 2);
        if (val_bits == 0) val_bits = 1;
        concat_bits = (concat_bits << val_bits) | val;
    }

    // Split into rows of bit_width
    mpz_class mask = (mpz_class(1) << bit_width) - 1;
    std::vector<mpz_class> rows;
    mpz_class remaining = concat_bits;

    while (remaining > 0) {
        rows.push_back(remaining & mask);
        remaining >>= bit_width;
    }
    if (rows.empty()) rows.push_back(0);

    // Fold: even rows XOR, odd rows ADD with reversed bits
    mpz_class result = 0;
    for (size_t i = 0; i < rows.size(); ++i) {
        if (i % 2 == 0) {
            result = result ^ rows[i];
        } else {
            mpz_class rev = reverse_bits(rows[i], bit_width);
            result = (result + rev) & mask;
        }
    }

    // Weighted sum mix
    mpz_class weighted_sum = 0;
    for (size_t i = 0; i < sequence.size(); ++i) {
        weighted_sum += sequence[i] * (i + 1);
    }
    result = (result ^ (weighted_sum & mask)) & mask;

    // To hex
    int target_len = bit_width / 4;
    std::string hex_hash = result.get_str(16);
    if ((int)hex_hash.size() < target_len) {
        hex_hash = std::string(target_len - hex_hash.size(), '0') + hex_hash;
    } else if ((int)hex_hash.size() > target_len) {
        hex_hash = hex_hash.substr(hex_hash.size() - target_len);
    }
    return hex_hash;
}

std::string collatz_hash_256(const std::string& text) {
    return collatz_hash(text, 256);
}

std::string collatz_hash_512(const std::string& text) {
    return collatz_hash(text, 512);
}

int main(int argc, char** argv) {
    if (argc < 2) {
        std::cout << "Usage: collatz_hash <text> [benchmark_iterations]\n";
        std::cout << "  With 1 arg: hash the text and print result\n";
        std::cout << "  With 2 args: benchmark N iterations\n\n";
        std::cout << "Examples:\n";
        std::cout << "  ./collatz_hash 'hello'\n";
        std::cout << "  ./collatz_hash 'password' 10000\n";
        return 1;
    }

    std::string text = argv[1];

    if (argc >= 3) {
        int iters = std::stoi(argv[2]);
        auto start = std::chrono::high_resolution_clock::now();
        for (int i = 0; i < iters; ++i) {
            collatz_hash_256(text);
        }
        auto end = std::chrono::high_resolution_clock::now();
        double ms = std::chrono::duration<double, std::milli>(end - start).count();
        std::cout << "C++ collatz_hash_256: " << (ms / iters) << " ms/op ("
                  << (iters / (ms / 1000.0)) << " ops/sec) over " << iters << " iterations\n";
    } else {
        std::string h256 = collatz_hash_256(text);
        std::string h512 = collatz_hash_512(text);
        std::cout << "Input: \"" << text << "\"\n";
        std::cout << "Hash 256: " << h256 << "\n";
        std::cout << "Hash 512: " << h512.substr(0, 64) << "...\n";
    }
    return 0;
}
